### DISCUSSION TITLE
Invoke OpenModelica Compiler from a C/C++ program

### DISCUSSION THREAD

**Discussion by kbenne** [OpenModelica/OpenModelica#10802] [Q&A] [answered]
This might seem unconventional, but I would like to embed the OpenModelica Compiiler into a domain specific C++ main program and invoke the compiler programatically. Can someone point me to a C/C++ API within OM that would allow me to compile a model? One option that seems appealing is to use the [scripting API]( https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/scripting_api.html), but if this is a good approach how do I invoke a script programatically? Is there a function along the lines of `void executeScript(...)` that would support my use case?

There seems to be a similar discussion thread [here](https://github.com/OpenModelica/OpenModelica/discussions/10185), but from what I can tell that discussion does not address the question of how to invoke a script from the C++ side.

**Comment by adeas31** [comment, marked as answer]
Look into the OMEdit source code. OMEdit links with the OpenModelica Compiler dll and uses the C api directly.

Alternative is to use the zeromq interface of OpenModelica Compiler. Look into the OMPython code for inspiration.

**Reply by kbenne** [reply]
Thank you. I think I've been sufficiently inspired by the OMEdit code. If I achieve a working prototype I'll be sure to link my example code back to this discussion in case anyone else tries to go down this path.

**Reply by casella** [reply]
Good. Maybe you can post some links here to GitHub displays of relevant chunks of code, for other to benefit.

**Reply by kbenne** [reply]
I think `omc_Main_handleCommand` is the heart of it, which we can see being utilized by OMEdit [here](https://github.com/OpenModelica/OpenModelica/blob/b8f9594377cc9852e222d20e41107468599d4e89/OMEdit/OMEditLIB/OMC/OMCProxy.cpp#L332). I'll provide more links if/when I have an example.

**Reply by kbenne** [reply]
I just wanted to follow up in case anyone else travels down the same road. Here is a minmal example of linking against OpenModelica and compiling a model using OM APIs. 

https://github.com/NREL/Spawn/blob/3c40cadf53709e92688f1de922b687451be21898/open_modelica/test/test_compiler.cpp

Thank you to the OM developers for making this possible.
